### DISCUSSION TITLE
Using map in another projection

### DISCUSSION THREAD

**Discussion by piotr-przebieracz** [openlayers/openlayers#14891] [Q&A] [answered]
Hello,

We have been using the OpenLayers library for quite some time now to display different types of maps including maps with tiles generated by ourselves as well as maps with GIS layers. So far we have only used GIS layers using default EPSG:3857 projection but now we would like to be able to use different projection types as well. Our question is: how should we properly display only the "correct/meaningful" part of the map while retaining the ability to locate other features based on their GPS location?

We saw the examples "Reprojection with EPSG.io search":
https://openlayers.org/en/latest/examples/reprojection-by-code.html
and "How do I change the projection of my map?” section of the FAQ:
https://openlayers.org/doc/faq.html
but it looks like it is not enough for some cases.

Let's assume we would like to use EPSG:28992 projection and display this NL Basic map using raster tiles from MapTiler:
https://cloud.maptiler.com/maps/basic-28992/

Initially we thought that if some tiles use specific projection (in this case EPSG:28992) it means that on the zoom level equal to zero the whole tile should contain only the extent of the projection (in this case: [3.2, 50.75, 7.22, 53.7] as WGS84 bounds):
![ghdisc01](https://github.com/openlayers/openlayers/assets/886251/b6bf948b-4aae-4869-af72-74bb5f885248)

However, as we can see in the provided map available on MapTiler that it is not the case - the tile at initial zoom level contains much bigger extent:
https://api.maptiler.com/maps/basic-28992/256/0/0/0.png?key=Bsv0SVuT0pvpxB218uly
![ghdisc02](https://github.com/openlayers/openlayers/assets/886251/690a1e20-7e42-4def-867d-1d015397c0b2)

Does it mean that two tiled maps using the same projection (e.g. EPSG:28992) can in fact represent a different area (bigger than specified as extent of the projection)?
How can we use such tiles in OpenLayers while retaining the ability to locate other features based on their GPS location?
We are wondering if there is any standard for obtaining such information for the set of tiles (e.g. extent of the whole tile grid, resolutions for the tile grid)? 

We tried to prepare a sample which shows that without additional information we are unable to correctly show this tile layer with ability to locate other features based on their GPS location:
https://codesandbox.io/s/using-epsg-28992-projection-simple-nxr6zt?file=/main.js
However, by checking the source code of the sample prepared by MapTiler we found some additional information which made the whole sample working:
https://codesandbox.io/s/using-epsg-28992-projection-with-all-data-v4p6rw?file=/main.js

The question is: how (and if) can we get the extent of the initial tile (and possible other information like available resolutions) based only on the provided spatial reference number?

**Comment by ahocevar** [comment]
This is more a question for Maptiler than for OpenLayers, because the answer is Maptiler specific. But anyway, I'll do my best:

They use an extended TileJSON format to advertise the parameters you are looking for. For the specific map, you can find the TileJSON here: https://api.maptiler.com/tiles/v3-28992/tiles.json?key=Bsv0SVuT0pvpxB218uly.

You'll want to look at the properties `extent` and `tile_matrix_set`. To get the resolution for each zoom level, you can simply use the `pixel_x_size` of each of the `items`.

You can find that `tiles.json` file referenced in `style.json` for the vector tile version of the tile set, which is referenced on the [page you linked](https://cloud.maptiler.com/maps/basic-28992/) under "Use vector style":

<img width="669" alt="Untitled" src="https://github.com/openlayers/openlayers/assets/211514/543d180d-15f3-4fff-acc0-a53538398711">

In this case: https://api.maptiler.com/maps/basic-28992/style.json?key=Bsv0SVuT0pvpxB218uly, which contains:
```json
"sources": {
  "openmaptiles": {
    "url": "https://api.maptiler.com/tiles/v3-28992/tiles.json?key=Bsv0SVuT0pvpxB218uly",
    "type":"vector"
  }
}
```

**Comment by piotr-przebieracz** [comment]
First of all, I would like to thank you very much for fast response.
About this particular map - you're right. I was rather focused on using raster tiles and links for 256x256 and 512x512 tiles didn't provide such information, e.g. response for: https://api.maptiler.com/maps/basic-28992/256/tiles.json?key=Bsv0SVuT0pvpxB218uly
contains only following data (with bounds strangely pointing even to EPSG:3857 projection):
```
{
"attribution":"<a href=\"https://www.maptiler.com/copyright/\" target=\"_blank\">&copy; MapTiler</a> <a href=\"https://www.openstreetmap.org/copyright\" target=\"_blank\">&copy; OpenStreetMap contributors</a>",
"tilejson":"2.0.0",
"name":"NL Basic (EPSG:28992)",
"minzoom":0,
"maxzoom":14,
"bounds":[-180,-85.0511,180,85.0511],
"format":"png",
"type":"baselayer",
"center":[4.766,52.187,1],
"color":"hsl(47, 26%, 88%)",
"tiles":["https://api.maptiler.com/maps/basic-28992/256/{z}/{x}/{y}.png?key=Bsv0SVuT0pvpxB218uly"],
"logo":"https://api.maptiler.com/resources/logo.svg"
}
``` 
However, as you rightly pointed out these parameters were available for vector tiles.

Anyway, maybe I didn't express myself clearly enough because my point was to find out common approach in such situations (based on a provided example). I understand that TileJSON is one of the formats for these parameters/metadata but what about situations in which this format is not supported by the map creator? I mean: do two tiled maps using the same projection (e.g. EPSG:28992) can in fact represent a different area (bigger than specified as extent of the projection)? Does it mean that the map creator should always provide these information in some way (e.g. TileJSON format or another specification)?

**Reply by ahocevar** [reply, marked as answer]
Like I said, my answer was Maptiler specific, and for details you'll have to contact them.

There are standards for what you're asking for (recently OGC API Tiles, perviously WMTS), but I don't know if Maptiler provides metadata according to these standards. The problem with tile sets in projections other than Web Mercator (EPSG:3857) that you need a tile grid definition, which needs an origin, which does not come naturally for local projections. I prefer to use the validity extent as defined in the EPSG database, but this is a decision of the provider of the map tiles. 

When following the mentioned standards, this information is contained in the metadata, and OpenLayers can work with the metadata directly - OpenLayers supports both WMTS and OGC API Tiles.

For WMTS, you can take a look at the [WMTS examples](https://openlayers.org/en/latest/examples/?q=wmts).
For OGC API Tiles, we also have an [example](https://openlayers.org/en/latest/examples/ogc-map-tiles.html).

**Reply by piotr-przebieracz** [reply]
Thank you very much for your help and support. It is now clear that if we use XYZ as the source for the layer we have to provide these information by ourselves (based on metadata which can be different depending on the tiles server like MapTiler, ArcGIS, etc.) or we can use specific source types like the examples you provided.
